crates/rspack_napi/src/threadsafe_js_value_ref.rs

```diff
@@ -0,0 +1,149 @@
+use std::ffi::{c_void, CString};
+use std::marker::PhantomData;
+use std::ptr;
+use std::sync::{Arc, Mutex};
+
+use napi::bindgen_prelude::*;
+use napi::sys::napi_threadsafe_function;
+use napi::NapiValue;
+
+use crate::js_values::js_value_ref::JsValueRef;
+
+pub struct ThreadsafeJsValueRef<T: NapiValue> {
+  inner: Arc<(Mutex<JsValueRef<T>>, DropJsValueRefFn<T>)>,
+}
+
+unsafe impl<T: NapiValue> Send for ThreadsafeJsValueRef<T> {}
+unsafe impl<T: NapiValue> Sync for ThreadsafeJsValueRef<T> {}
+
+impl<T: NapiValue> Clone for ThreadsafeJsValueRef<T> {
+  fn clone(&self) -> Self {
+    Self {
+      inner: self.inner.clone(),
+    }
+  }
+}
+
+impl<T: NapiValue> Drop for ThreadsafeJsValueRef<T> {
+  fn drop(&mut self) {
+    if Arc::strong_count(&self.inner) == 1 {
```

Comment:
It's not sound here. `Arc::strong_count` uses `Relax` ordering. You can leverage `Arc`'s drop when dropping the inner content. For example: `Arc<NewType<Mutex<...>>>` and implement `fn drop` for `NewType`.